### DISCUSSION TITLE
Aud Claim missing in ID Token

### DISCUSSION THREAD

**Discussion by MagiiEl** [keycloak/keycloak#22520] [Q&A] [answered]
Hi,
while im researching about the openID Connect Standard and how to validate ID and Access Tokens properly. I am assuming that I have to check the aud claim in the ID Token sent to the frontend which should only be my client.

Now I am stuck because there is no "aud" claim set in the ID Token during login. But If I'm evaluating the client scopes in the admin console (clients > client > Client scopes > evaluate), the generated ID Token has the correct claim.

I feel like this is a bug? the [openid connect specs](https://openid.net/specs/openid-connect-core-1_0.html#HybridIDToken2) also define that claim as required:
![grafik](https://github.com/keycloak/keycloak/assets/13433812/d057a45b-1ae2-42bc-aa01-4ebc8ba3af64)

How can I make sure that the audience claim is in the ID Token without having to hard-code something? Or what are my option? I do want to make the validation process as close to the standard as possible

**Comment by justin-tay** [comment]
From what I see by default there should be a `aud` claim and a `azp` claim with the same value in the ID Token by default. The Access Token though doesn't have the `aud` of the `client_id` set by default.

Maybe you can elaborate a bit about your setup, for instance which version of Keycloak you are using.

You can add additional claims by adding mappers. In `Client scopes` you can click on the `Dedicated scope and mappers for this client`. You can hardcode an additional `aud` by `Add mapper` > `By configuration` > `Audience`. You can then select whether you want to `Add to ID token` or `Add to access token`.

**Reply by MagiiEl** [reply]
Thank you for your fast reply. I am using Keycloak 21.1.2.  The Clients are normal standard flows. We didnt really touch any scope mapper but just adding a custom claim to the access token. The ID Token that gets issued looks like this (removed explicit information). I know see that the typ is not simply "ID" but "Serialized-ID". I dont know if that makes a difference.:

```
{
  "exp": 1692393389,
  "iat": 1692357389,
  "jti": ...,
  "iss": "...",
  "sub": "...",
  "typ": "Serialized-ID",
  "session_state": "..",
  "sid": "...",
  "state_checker": "..."
}

**Reply by justin-tay** [reply]
How are you getting this token value? This looks like the value in the `KEYCLOAK_IDENTITY` cookie.

Are you getting the `id_token` from the token endpoint?

![id-token](https://github.com/keycloak/keycloak/assets/49700559/10afe7b3-c6d3-4cc1-8619-2a4db4424193)

The decoded value of this `id_token` is

```
{
  "exp": 1692360476,
  "iat": 1692360176,
  "auth_time": 0,
  "jti": "35a3f12c-b3cd-4af5-9930-b6ae0b161bc3",
  "iss": "http://localhost:8080/realms/test",
  "aud": "test",
  "sub": "47d77795-25b2-4c5e-a044-2b4b06ffcc30",
  "typ": "ID",
  "azp": "test",
  "session_state": "b8208ed9-ad27-460b-9ce8-f11624a286f1",
  "at_hash": "qZhFgjprVY3uYWlA3frbLw",
  "acr": "1",
  "sid": "b8208ed9-ad27-460b-9ce8-f11624a286f1",
  "email_verified": false,
  "preferred_username": "testuser",
  "given_name": "",
  "family_name": ""
}
```

**Reply by MagiiEl** [reply]
Yes, I was looking at the Keycloak ID as I assumed that the ID Token will be sent directly to the browser via standard flow. It seems like that is not the case, then? I would have to use the direct-grant flow (have it activated) to get proper ID token with that request via my backend receives the the access token?

**Reply by justin-tay** [reply]
The relying party will be able to get the `id_token`  via standard flow at the token endpoint as well at the same time it gets the `access_token` and `refresh_token`. In your case is your frontend the relying party (public client) or is the backend the relying party (confidential client)? It's not really clear what you are trying to do. If your frontend is not a relying party then it doesn't have to deal with the tokens at all.

**Reply by MagiiEl** [reply]
These are the settings of the client:
![grafik](https://github.com/keycloak/keycloak/assets/13433812/07248a23-2979-4a07-b046-713c7582355c)

I have a client ID and secret as well. 

But I can see know that I have been misunderstanding and actually mixing different information and coming to a false conclusion.

**The correct way of thinking would be something like this, I think:** 
The Relying Party receives *all* 3 tokens (access,id, refresh), which is essentially the backend part of the client. The frontend  (browser part) only receives the auth code from Keycloak upon successfull login which will be redirected to the backend. The backend can then receive the tokens by trading the auth code ,client secret (as we are a confidential client) and client id. I didn't check if the ID Token gets send automatically through that trade. But as you said you can retrieve it through the token endpoint. I would have to validate all tokens in the backend before processing the login.

The Keycloak Cookies are there for SSO, so it makes sense it isnt a classic ID Token as it could technically stand for several "Relying Parties". 

Thank you for taking your time answering and helping me through my confusion.

**Reply by justin-tay** [reply, marked as answer]
To get the `id_token` at the token endpoint the scope must include `openid`.

That is correct for a browser frontend where the backend is the confidential client. The backend is the one to get the tokens from Keycloak. The backend maintains a session with the browser frontend using its own `HttpOnly` `Secure` session cookie. 

For native mobile application frontends, it can be a public client and perform the standard flow with PKCE using AppAuth. In that case the native mobile frontend is the one to get the tokens. It can use the `id_token` as a login credential. The `azp` is the client id representing the mobile frontend and the `aud` should be the backend client id that the token is presented to for login.

**Reply by MagiiEl** [reply]
Thank you a lot for helping me clarify!
